File: packages/react-native-hms/android/src/main/java/com/reactnativehmssdk/Interactivity/HMSInteractivityDecoder.kt
package com.reactnativehmssdk.Interactivity

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.reactnativehmssdk.HMSDecoder
import live.hms.video.polls.models.*
import live.hms.video.polls.models.answer.HMSPollQuestionAnswer
import live.hms.video.polls.models.answer.HmsPollAnswer
import live.hms.video.polls.models.answer.PollAnswerItem
import live.hms.video.polls.models.answer.PollAnswerResponse
import live.hms.video.polls.models.network.HMSPollResponsePeerInfo
import live.hms.video.polls.models.question.HMSPollQuestion
import live.hms.video.polls.models.question.HMSPollQuestionOption
import live.hms.video.polls.network.HMSPollLeaderboardEntry
import live.hms.video.polls.network.HMSPollLeaderboardSummary
import live.hms.video.polls.network.PollLeaderboardResponse
import live.hms.video.polls.network.PollResultsDisplay

object HMSInteractivityDecoder {
  fun getPollUpdateType(hmsPollUpdateType: HMSPollUpdateType): Int {
    return when (hmsPollUpdateType) {
      HMSPollUpdateType.started -> 0
      HMSPollUpdateType.resultsupdated -> 1
      HMSPollUpdateType.stopped -> 2
      else -> {
        0
      }
    }
  }

  fun getPoll(poll: HmsPoll): WritableMap {
    val data = Arguments.createMap()

    data.putBoolean("anonymous", poll.anonymous)
    data.putInt("type", poll.category.ordinal)
    poll.duration?.let { data.putDouble("duration", it.toDouble()) }
    data.putString("pollId", poll.pollId)
    data.putArray("rolesThatCanViewResponses", HMSDecoder.getAllRoles(poll.rolesThatCanViewResponses))
    data.putArray("rolesThatCanVote", HMSDecoder.getAllRoles(poll.rolesThatCanVote))
    data.putInt("state", getPollStateOrdinal(poll.state))
    data.putString("title", poll.title)

    poll.createdBy?.let {
      data.putMap("createdBy", HMSDecoder.getHmsPeerSubset(it))
    }

    poll.mode?.let {
      data.putInt("mode", getPollMode(poll))
    }

    poll.questionCount?.let {
      data.putInt("questionCount", it)
    }

    poll.questions?.let {
      data.putArray("questions", getPollQuestions(it))
    }

    poll.result?.let {
      data.putMap("result", getPollResult(it))
    }

    data.putString("startedAt", poll.startedAt.toString())

    poll.stoppedAt?.let {
      data.putString("stoppedAt", it.toString())
    }

    poll.startedBy?.let {
      data.putMap("startedBy", HMSDecoder.getHmsPeerSubset(it))
    }

    return data
  }

  private fun getPollStateOrdinal(pollState: HmsPollState): Int {
    return when (pollState) {
      HmsPollState.CREATED -> 0
      HmsPollState.STARTED -> 1
      HmsPollState.STOPPED -> 2
      else -> 0
    }
  }

  private fun getPollMode(poll: HmsPoll): Int {
    return when (poll.mode) {
      HmsPollUserTrackingMode.PEER_ID -> 0
      HmsPollUserTrackingMode.USER_ID -> 1
      HmsPollUserTrackingMode.USERNAME -> 2
      else -> {
        0
      }
    }
  }

  private fun getPollQuestions(questions: List<HMSPollQuestion>): WritableArray {
    val data = Arguments.createArray()

    for (question in questions) {
      data.pushMap(getPollQuestion(question))
    }

    return data
  }

  private fun getPollQuestion(question: HMSPollQuestion): WritableMap {
    val data = Arguments.createMap()

    data.putInt("index", question.questionID)

    question.duration.let { data.putDouble("duration", it.toDouble()) }

    question.myResponses.let {
      data.putArray("myResponses", getPollAnswers(it))
    }

    data.putBoolean("skippable", question.canSkip)

    data.putString("text", question.text)

    data.putInt("type", question.type.ordinal)

    data.putBoolean("voted", question.voted)

    data.putInt("weight", question.weight)

    question.correctAnswer?.let {
      data.putMap("answer", getPollQuestionAnswer(it))
    }

    question.answerLongMinLength?.let {
      data.putDouble("answerMaxLen", it.toDouble())
    }

    question.answerShortMinLength?.let {
      data.putDouble("answerMinLen", it.toDouble())
    }

    question.options?.let {
      data.putArray("options", getPollQuestionOptions(it))
    }

    return data
  }

  private fun getPollQuestionOptions(options: List<HMSPollQuestionOption>): WritableArray {
    val data = Arguments.createArray()

    for (option in options) {
      data.pushMap(getPollQuestionOption(option))
    }

    return data
  }

  private fun getPollQuestionOption(option: HMSPollQuestionOption): WritableMap {
    val data = Arguments.createMap()

    data.putString("text", option.text)
    data.putInt("index", option.index)
    data.putInt("voteCount", option.voteCount.toInt())
    option.weight?.let { data.putInt("weight", it) }

    return data
  }

  private fun getPollQuestionAnswer(answer: HMSPollQuestionAnswer): WritableMap {
    val data = Arguments.createMap()

    answer.option?.let { data.putInt("option", it) }

    answer.options?.let { data.putArray("options", convertListToReadableArray(it)) }

    data.putBoolean("hidden", answer.hidden)

    return data
  }

  private fun getPollAnswers(answers: List<HmsPollAnswer>): WritableArray {
    val data = Arguments.createArray()

    for (answer in answers) {
      data.pushMap(getPollAnswer(answer))
    }

    return data
  }

  private fun getPollAnswer(answer: HmsPollAnswer): WritableMap {
    val data = Arguments.createMap()

    answer.durationMillis.let { data.putDouble("duration", (it * 1000).toDouble()) }

    data.putInt("option", answer.selectedOption)

    data.putInt("questionId", answer.questionId)

    data.putBoolean("skipped", answer.skipped)

    data.putString("text", answer.answerText)

    data.putInt("type", answer.questionType.ordinal)

    data.putBoolean("update", answer.update)

    answer.selectedOptions?.let {
      data.putArray("options", convertListToReadableArray(it))
    }

    return data
  }

  private fun getPollResult(result: PollResultsDisplay): WritableMap {
    val data = Arguments.createMap()

    result.totalResponses?.let {
      data.putDouble("totalResponse", it.toDouble())
    }

    result.votingUsers?.let {
      data.putDouble("userCount", it.toDouble())
    }

    result.totalDistinctUsers?.let {
      data.putDouble("maxUserCount", it.toDouble())
    }

    result.questions.let {
      data.putArray("questions", getPollStatsQuestion(it))
    }

    return data
  }

  private fun getPollStatsQuestion(questions: List<PollStatsQuestions>): WritableArray {
    val data = Arguments.createArray()

    for (question in questions) {
      data.pushMap(getPollStatsQuestion(question))
    }

    return data
  }

  private fun getPollStatsQuestion(question: PollStatsQuestions): WritableMap {
    val data = Arguments.createMap()

    data.putInt("question", question.index)

    data.putInt("type", question.questionType.ordinal)

    question.correct?.let { data.putDouble("correctVotes", it.toDouble()) }

    data.putDouble("skippedVotes", question.skipped.toDouble())

    data.putInt("totalVotes", question.attemptedTimes)

    question.options?.let {
      data.putArray("optionVoteCounts", convertLongListToReadableArray(it))
    }

    return data
  }

  private fun convertLongListToReadableArray(list: List<Long>): ReadableArray {
    val readableArray = Arguments.createArray()
    list.forEach {
      readableArray.pushInt(it.toInt())
    }
    return readableArray
  }

  private fun convertListToReadableArray(list: List<Int>): ReadableArray {
    val readableArray = Arguments.createArray()
    list.forEach {
      readableArray.pushInt(it.toInt())
    }
    return readableArray
  }

/*

  static func getHMSPollQuestionResponseResults(_ hmsPollQuestionResponseResults: [HMSPollQuestionResponseResult]) -> [[String: AnyHashable]] {
    var results = [[String: AnyHashable]]()

    hmsPollQuestionResponseResults.forEach { result in
      results.append(getHMSPollQuestionResponseResult(result))
    }
    return results
  }

  static func getHMSPollQuestionResponseResult(_ hmsPollQuestionResponseResult: HMSPollQuestionResponseResult) -> [String: AnyHashable] {
    var result: [String: AnyHashable] = [
    "question": hmsPollQuestionResponseResult.question
    ]
    if let correct = hmsPollQuestionResponseResult.correct {
      result["correct"] = correct
    }
    if let error = hmsPollQuestionResponseResult.error {
      result["error"] = error.localizedDescription
    }
    return result
  }

 */
  fun getHMSPollQuestionResponseResults(hmsPollQuestionResponseResults: PollAnswerResponse): WritableArray {
    val results = Arguments.createArray()

    hmsPollQuestionResponseResults.result.forEach { result ->
      results.pushMap(getHMSPollQuestionResponseResult(result))
    }
    return results
  }

  fun getPollLeaderboardResponse(pollLeaderboardResponse: PollLeaderboardResponse): WritableMap {
    val results = Arguments.createMap()

    pollLeaderboardResponse.hasNext?.let {
      results.putBoolean("hasNext", it)
    }
    pollLeaderboardResponse.summary?.let {
      results.putMap("summary", getHMSPollLeaderboardSummary(it))
    }
    pollLeaderboardResponse.entries?.let {
      results.putArray("entries", getHMSPollLeaderboardEntries(it))
    }
    return results
  }

  private fun getHMSPollLeaderboardSummary(pollLeaderboardSummary: HMSPollLeaderboardSummary): WritableMap {
    val summary = Arguments.createMap()
    pollLeaderboardSummary.averageScore?.let {
      summary.putDouble("averageScore", it.toDouble())
    }
    pollLeaderboardSummary.averageTime?.let {
      summary.putString("averageTime", it.toString())
    }
    pollLeaderboardSummary.totalPeersCount?.let {
      summary.putInt("totalPeersCount", it)
    }
    pollLeaderboardSummary.respondedCorrectlyPeersCount?.let {
      summary.putInt("respondedCorrectlyPeersCount", it)
    }
    pollLeaderboardSummary.respondedPeersCount?.let {
      summary.putInt("respondedPeersCount", it)
    }
    return summary
  }

  private fun getHMSPollLeaderboardEntries(pollLeaderboardEntries: List<HMSPollLeaderboardEntry>): WritableArray {
    val list = Arguments.createArray()
    pollLeaderboardEntries.forEach {
      list.pushMap(getHMSPollLeaderboardEntry(it))
    }
    return list
  }

  private fun getHMSPollLeaderboardEntry(pollLeaderboardEntry: HMSPollLeaderboardEntry): WritableMap {
    val entry = Arguments.createMap()
    pollLeaderboardEntry.duration?.let {
      entry.putString("duration", it.toString())
    }
    pollLeaderboardEntry.peer?.let {
      entry.putMap("peer", getHMSPollResponsePeerInfo(it))
    }
    pollLeaderboardEntry.totalResponses?.let {
      entry.putString("totalResponses", it.toString())
    }
    pollLeaderboardEntry.correctResponses?.let {
      entry.putString("correctResponses", it.toString())
    }
    pollLeaderboardEntry.position?.let {
      entry.putString("position", it.toString())
    }
    pollLeaderboardEntry.score?.let {
      entry.putString("score", it.toString())
    }
    return entry
  }

  private fun getHMSPollResponsePeerInfo(pollResponsePeerInfo: HMSPollResponsePeerInfo): WritableMap {
    val peerInfo = Arguments.createMap()
    peerInfo.putString("userHash", pollResponsePeerInfo.hash)
    pollResponsePeerInfo.peerid?.let {
      peerInfo.putString("peerId", it)
    }
    pollResponsePeerInfo.userid?.let {
      peerInfo.putString("customerUserId", it)
    }
    pollResponsePeerInfo.username?.let {
      peerInfo.putString("userName", it)
    }
    return peerInfo
  }

  private fun getHMSPollQuestionResponseResult(hmsPollQuestionResponseResult: PollAnswerItem): WritableMap {
    val result = Arguments.createMap()

    result.putInt("question", hmsPollQuestionResponseResult.questionIndex)

    hmsPollQuestionResponseResult.correct.let {
      result.putBoolean("correct", it)
    }

    hmsPollQuestionResponseResult.error?.let {
      result.putString("error", it.localizedMessage)
    }

    return result
  }
}

File: packages/react-native-hms/ios/HMSInteractivityDecoder.swift
//
//  HMSInteractivityDecoder.swift
//  react-native-hms
//
//  Created by Jatin Nagar on 18/01/24.
//

import Foundation
import HMSSDK

class HMSInteractivityDecoder {
    static func getHMSPoll(_ poll: HMSPoll) -> [String: Any] {
        var dict: [String: Any] = [
            "anonymous": poll.anonymous,
            "type": poll.category.rawValue,
            "duration": poll.duration,
            "pollId": poll.pollID,
            "rolesThatCanViewResponses": HMSDecoder.getAllRoles(poll.rolesThatCanViewResponses),
            "rolesThatCanVote": HMSDecoder.getAllRoles(poll.rolesThatCanVote),
            "state": poll.state.rawValue,
            "title": poll.title
        ]
        if let createdBy = poll.createdBy {
            dict["createdBy"] = HMSDecoder.getHmsPeerSubset(createdBy)
        }
        if let mode = poll.mode {
            dict["mode"] = mode.rawValue
        }
        if let questionCount = poll.questionCount {
            dict["questionCount"] = questionCount
        }
        if let questions = poll.questions {
            dict["questions"] = getHMSPollQuestions(questions)
        }
        if let result = poll.result {
            dict["result"] = getHMSPollResult(result)
        }
        if let startedAt = poll.startedAt {
            dict["startedAt"] = startedAt.timeIntervalSince1970 * 1000
        }
        if let startedBy = poll.startedBy {
            dict["startedBy"] = HMSDecoder.getHmsPeerSubset(startedBy)
        }
        if let stoppedAt = poll.stoppedAt {
            dict["stoppedAt"] = stoppedAt.timeIntervalSince1970 * 1000
        }
        if let stoppedBy = poll.stoppedBy {
            dict["stoppedBy"] = HMSDecoder.getHmsPeerSubset(stoppedBy)
        }
        return dict
    }

    static func getHMSPollQuestions(_ hmsPollQuestions: [HMSPollQuestion]) -> [[String: AnyHashable]] {
        var questions = [[String: AnyHashable]]()

        hmsPollQuestions.forEach { question in
            questions.append(getHMSPollQuestion(question))
        }
        return questions
    }

    static func getHMSPollQuestion(_ hmsPollQuestion: HMSPollQuestion) -> [String: AnyHashable] {
        var dict: [String: AnyHashable] = [
            "duration": hmsPollQuestion.duration,
            "index": hmsPollQuestion.index,
            "myResponses": getHMSPollQuestionResponses(hmsPollQuestion.myResponses),
            "once": hmsPollQuestion.once,
            "skippable": hmsPollQuestion.skippable,
            "text": hmsPollQuestion.text,
            "type": hmsPollQuestion.type.rawValue,
            "voted": hmsPollQuestion.voted,
            "weight": hmsPollQuestion.weight
        ]
        if let answer = hmsPollQuestion.answer {
            dict["answer"] = getHMSPollQuestionAnswer(answer)
        }
        if let answerMaxLen = hmsPollQuestion.answerMaxLen {
            dict["answerMaxLen"] = answerMaxLen
        }
        if let answerMinLen = hmsPollQuestion.answerMinLen {
            dict["answerMinLen"] = answerMinLen
        }
        if let options = hmsPollQuestion.options {
            dict["options"] = getHMSPollQuestionOptions(options)
        }
        if let responses = hmsPollQuestion.responses {
            dict["responses"] = getHMSPollQuestionResponses(responses)
        }
        return dict
    }

    static func getHMSPollQuestionAnswer(_ hmsPollQuestionAnswer: HMSPollQuestionAnswer) -> [String: AnyHashable] {
        var dict: [String: AnyHashable] = [
            "hidden": hmsPollQuestionAnswer.hidden
        ]
        if let option = hmsPollQuestionAnswer.option {
            dict["option"] = option
        }
        if let options = hmsPollQuestionAnswer.options {
            dict["options"] = options
        }
        return dict
    }

    static func getHMSPollQuestionOptions(_ hmsPollQuestionOptions: [HMSPollQuestionOption]) -> [[String: AnyHashable]] {
        var dict = [[String: AnyHashable]]()
        hmsPollQuestionOptions.forEach { hmsPollQuestionOption in
            dict.append(getHMSPollQuestionOption(hmsPollQuestionOption))
        }
        return dict
    }

    static func getHMSPollQuestionOption(_ hmsPollQuestionOption: HMSPollQuestionOption) -> [String: AnyHashable] {
        return [
            "text": hmsPollQuestionOption.text,
            "index": hmsPollQuestionOption.index,
            "voteCount": hmsPollQuestionOption.voteCount,
            "weight": hmsPollQuestionOption.weight
        ]
    }

    static func getHMSPollQuestionResponses(_ hmsPollQuestionResponses: [HMSPollQuestionResponse]) -> [[String: AnyHashable]] {
        var dict = [[String: AnyHashable]]()
        hmsPollQuestionResponses.forEach { hmsPollQuestionResponse in
            dict.append(getHMSPollQuestionResponse(hmsPollQuestionResponse))
        }
        return dict
    }

    static func getHMSPollQuestionResponse(_ hmsPollQuestionResponse: HMSPollQuestionResponse) -> [String: AnyHashable] {
        var dict: [String: AnyHashable] = [
            "duration": hmsPollQuestionResponse.duration,
            "option": hmsPollQuestionResponse.option,
            "questionId": hmsPollQuestionResponse.questionID,
            "skipped": hmsPollQuestionResponse.skipped,
            "text": hmsPollQuestionResponse.text,
            "type": hmsPollQuestionResponse.type.rawValue,
            "update": hmsPollQuestionResponse.update
        ]
        if let options = hmsPollQuestionResponse.options {
            dict["options"] = options
        }
        if let peerInfo = hmsPollQuestionResponse.peer {
            dict["peer"] = getHMSPollResponsePeerInfo(peerInfo)
        }
        if let responseFinal = hmsPollQuestionResponse.responseFinal {
            dict["responseFinal"] = responseFinal
        }
        return dict
    }

    static func getHMSPollResponsePeerInfo(_ hmsPollResponsePeerInfo: HMSPollResponsePeerInfo) -> [String: AnyHashable] {
        var dict = [String: AnyHashable]()

        if let customerUserID = hmsPollResponsePeerInfo.customerUserID {
            dict["customerUserId"] = customerUserID
        }
        if let peerID = hmsPollResponsePeerInfo.peerID {
            dict["peerId"] = peerID
        }
        if let userHash = hmsPollResponsePeerInfo.userHash {
            dict["userHash"] = userHash
        }
        if let userName = hmsPollResponsePeerInfo.userName {
            dict["userName"] = userName
        }
        return dict
    }

    static func getHMSPollResult(_ hmsPollResult: HMSPollResult) -> [String: AnyHashable] {
        return [
            "maxUserCount": hmsPollResult.maxUserCount,
            "questions": getHMSPollQuestionResults(hmsPollResult.questions),
            "totalResponse": hmsPollResult.totalResponse,
            "userCount": hmsPollResult.userCount
        ]
    }

    static func getHMSPollQuestionResults(_ hmsPollQuestionResults: [HMSPollQuestionResult]) -> [[String: AnyHashable]] {
        var results = [[String: AnyHashable]]()

        hmsPollQuestionResults.forEach { result in
            results.append(getHMSPollQuestionResult(result))
        }
        return results
    }

    static func getHMSPollQuestionResult(_ hmsPollQuestionResult: HMSPollQuestionResult) -> [String: AnyHashable] {
        var result: [String: AnyHashable] = [
            "type": hmsPollQuestionResult.type,
            "correctVotes": hmsPollQuestionResult.correctVotes,
            "question": hmsPollQuestionResult.question,
            "skippedVotes": hmsPollQuestionResult.skippedVotes,
            "totalVotes": hmsPollQuestionResult.totalVotes
        ]
        if let optionVoteCounts = hmsPollQuestionResult.optionVoteCounts {
            result["optionVoteCounts"] = optionVoteCounts
        }
        return result
    }

    static func getHMSPollQuestionResponseResults(_ hmsPollQuestionResponseResults: [HMSPollQuestionResponseResult]) -> [[String: AnyHashable]] {
        var results = [[String: AnyHashable]]()

        hmsPollQuestionResponseResults.forEach { result in
            results.append(getHMSPollQuestionResponseResult(result))
        }
        return results
    }

    static func getHMSPollQuestionResponseResult(_ hmsPollQuestionResponseResult: HMSPollQuestionResponseResult) -> [String: AnyHashable] {
        var result: [String: AnyHashable] = [
            "question": hmsPollQuestionResponseResult.question
        ]
        if let correct = hmsPollQuestionResponseResult.correct {
            result["correct"] = correct
        }
        if let error = hmsPollQuestionResponseResult.error {
            result["error"] = error.localizedDescription
        }
        return result
    }

    static func getLeaderboardResponse(_ response: HMSPollLeaderboardResponse) -> [String: AnyHashable] {
        var result = [String: AnyHashable]()

        result["hasNext"] = response.hasNext

        result["summary"] = getLeaderboardSummary(response.summary)

        result["entries"] = getLeaderboardEntries(response.entries)

        return result
    }

    static func getLeaderboardSummary(_ summary: HMSPollLeaderboardSummary) -> [String: AnyHashable] {
        var result = [String: AnyHashable]()

        result["averageScore"] = summary.averageScore
        result["averageTime"] = summary.averageTime
        result["totalPeersCount"] = summary.totalPeersCount
        result["respondedCorrectlyPeersCount"] = summary.respondedCorrectlyPeersCount
        result["respondedPeersCount"] = summary.respondedPeersCount

        return result
    }

    static func getLeaderboardEntries(_ entries: [HMSPollLeaderboardEntry]) -> [[String: AnyHashable]] {
        var result = [[String: AnyHashable]]()

        for entry in entries {
            result.append(getLeaderboardEntry(entry))
        }

        return result
    }

    static func getLeaderboardEntry(_ entry: HMSPollLeaderboardEntry) -> [String: AnyHashable] {
        var result = [String: AnyHashable]()

        result["duration"] = entry.duration
        result["totalResponses"] = entry.totalResponses
        result["correctResponses"] = entry.correctResponses
        result["position"] = entry.position
        result["score"] = entry.score

        if let peer = entry.peer {
            result["peer"] = getHMSPollResponsePeerInfo(peer)
        }

        return result
    }

    static func getHMSWhiteboard(_ hmsWhiteboard: HMSWhiteboard) -> [String: Any] {
        var result: [String: Any] = [
            "id": hmsWhiteboard.id,
        ]
        if let owner = hmsWhiteboard.owner {
            result["owner"] = HMSDecoder.getHmsPeerSubset(owner)
        }
        if let title = hmsWhiteboard.title {
            result["title"] = title
        }
        if let urlString = hmsWhiteboard.url?.absoluteString {
            result["url"] = urlString
        }
        return result
    }

    static func getWhiteboardState(_ state: HMSWhiteboard.WhiteboardState) -> String {
        switch state {
        case .started:
            return "STARTED"
        case .stopped:
            return "STOPPED"
        @unknown default:
            print("@unknown default case: Invalid HMSWhiteboardUpdateType")
            return "STOPPED"
        }
    }

    static func getWhiteboardUpdateType(_ type: HMSWhiteboardUpdateType) -> String {
        switch type {
        case .started:
            return "STARTED"
        case .stopped:
            return "STOPPED"
        @unknown default:
            print("@unknown default case: Invalid HMSWhiteboardUpdateType")
            return "STOPPED"
        }
    }
}
